### ISSUE TITLE
Project level 9 test case 9_13 infinite looping

### ISSUE THREAD

**Issue by heeeyi** [nus-cs2030/2122-s2#392]
Hi my code works fine for all the test cases till level 8, however, it seem to enter infinite loop in level 9, test case 9_13... It is a hidden test case, so I do not even know where exactly goes wrong. I write all my Events as separate classes, and used Wait and PendingServe classes. I guess this might be an issue with these 2 classes, but I cannot pinpoint the issue. Does anyone ever encountered similar issues, or could you provide some suggestions about what might be the cause? Thx!

![屏幕截图(1142)](https://user-images.githubusercontent.com/97591125/163927712-9634e414-5541-44ea-85d5-8f5db2ed3394.png)

**Comment by weihan10** [comment]
I strongly suspect it's the PendingServe Event, I had a few friends encounter the same issue when the tried to create additional and unnecessary classes. I think you can try to examine from there.

**Comment by heeeyi** [comment]
Yeah I know it is probably PendingServe that goes wrong... Do you mean that PendingServe is an unnecessary class? Because I use it as a connection between Wait and Serve, I thought many people also do the same...

**Comment by weihan10** [comment]
> Yeah I know it is probably PendingServe that goes wrong... Do you mean that PendingServe is an unnecessary class? Because I use it as a connection between Wait and Serve, I thought many people also do the same...

It's not really unnecessary per se, but you can easily get around it with few lines of code and design. I chose to avoid it as it adds unnecessary complications to otherwise simple code. 
But since you have used it, you can continue to use it definitely no issue. Maybe you can try tracing the events one by one to find where the logic gap is. That's what I do to debug in my case.

**Comment by SoSimplySean** [comment]
Hi, experienced a few infinite loops as I was working through the last few levels - root cause generally fell to a variant of your PendingServe event as mentioned. Usually due to PendingServe continuing to call PendingServe, falling into a loop. Can try printing out your PQs and see whether they are being updated correctly. One of my errors was due to enqueuing a customer when I shouldn't have.

**Comment by heeeyi** [comment]
Hi just to help if anyone meet the same issue... Mine issue is probably due to a suboptimal class design. I do not fetch the fastest-ending self counters at the very start, but put it in one of the if-else block. This tests correctly on all the level 8 cases, but somehow goes wrong in level 9, which shows that it is not a good choice to fit edge cases. Putting it at the start in Wait and PendingServe run() helps.
